### PR TITLE
feat: simplify sha256 implementation

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Resolves 

## Summary\*

This unconstrained helper looks like it's taken a constrained implementation and not modified it to take advantage of the extra flexibility of the unconstrained runtime. I've then changed it to use a runtime for-loop to reduce the number of iterations.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: noir_stdlib/src/hash/sha256.nr [hunk 1]
```diff
@@ -34,39 +34,39 @@ fn msg_u8_to_u32(msg: [u8; 64]) -> [u32; 16] {
...
-    for k in msg_start..msg_end {
-        if k as u64 < message_size {
-            msg_block[msg_byte_ptr] = msg[k];
-            msg_byte_ptr = msg_byte_ptr + 1;
+    // We insert `BLOCK_SIZE` bytes (or up to the end of the message)
+    let block_input = if msg_start + BLOCK_SIZE > message_size {
+        if message_size > msg_start {
+            message_size - msg_start
+        } else {
+            0
```
Comment (TomAFrench) [inline noir_stdlib/src/hash/sha256.nr:44]:
```suggestion
        if msg_start < message_size  {
            // This function is sometimes called with `msg_start` past the end of the message
            // In this case we return an empty block and zero pointer to signal that the result should be ignored.
            0
        } else {
            message_size - msg_start
```

File: noir_stdlib/src/hash/sha256.nr [hunk 2]
```diff
@@ -34,39 +34,41 @@ fn msg_u8_to_u32(msg: [u8; 64]) -> [u32; 16] {
...
-        msg_end = N;
-    }
-    for k in msg_start..msg_end {
-        if k as u64 < message_size {
-            msg_block[msg_byte_ptr] = msg[k];
-            msg_byte_ptr = msg_byte_ptr + 1;
+    // We insert `BLOCK_SIZE` bytes (or up to the end of the message)
+    let block_input = if msg_start + BLOCK_SIZE > message_size {
+        if msg_start < message_size  {
+            // This function is sometimes called with `msg_start` past the end of the message
```
Comment (TomAFrench) [inline noir_stdlib/src/hash/sha256.nr:42]:
```suggestion
            // This function is sometimes called with `msg_start` past the end of the message.
```

File: noir_stdlib/src/hash/sha256.nr [hunk 3]
```diff
@@ -34,39 +34,41 @@ fn msg_u8_to_u32(msg: [u8; 64]) -> [u32; 16] {
...
-    if msg_end > N {
-        msg_end = N;
-    }
-    for k in msg_start..msg_end {
-        if k as u64 < message_size {
-            msg_block[msg_byte_ptr] = msg[k];
-            msg_byte_ptr = msg_byte_ptr + 1;
+    // We insert `BLOCK_SIZE` bytes (or up to the end of the message)
+    let block_input = if msg_start + BLOCK_SIZE > message_size {
+        if msg_start < message_size  {
```
Comment (TomAFrench) [inline noir_stdlib/src/hash/sha256.nr:41]:
```suggestion
        if message_size < msg_start  {
```

**Comment by TomAFrench** [comment]
Arggggggggghhhhhhhh. The formatter was a mistake.

**Review by vezenovm — APPROVED** [review]

**Pull request merged**